examples/async_replication/async_replicated_container.h

```diff
@@ -0,0 +1,421 @@
+  CURRENT_FIELD(clock, std::chrono::microseconds);
+};
+
+struct ReplicationNode final{
+  std::string host;
+  uint16_t port;
+};
+
+struct ReplicationConfig final {
+  std::string host;
+  uint16_t port;
+  std::vector<ReplicationNode> nodes_list;
+  uint32_t delay;
+  bool is_verbose;
+  bool show_network_errors;
+  uint32_t max_waits;
+};
+
+struct SharedState final {
+  bool die = false;
+  std::map<std::string, uint32_t > data;
+  std::map<std::string, std::chrono::microseconds > clock;
+  std::queue<std::pair<std::string, uint32_t> > replication_out;
+};
+
+
+class AsyncReplicatedContainer
+{
+ private:
+  std::string sid;
+  uint16_t reader_port;
+  bool is_ready = false;
+  std::vector<ReplicationNode> nodes;
+  current::WaitableAtomic<SharedState> state;
+  std::vector<std::thread> writers;
+  std::vector<std::thread> readers;
+  std::thread monitor;
+  uint32_t delay;
+  bool is_verbose;
+  bool show_network_err;
+  uint32_t max_waits;
+
+  std::string node_id(std::string host, uint16_t port){
+    std::ostringstream r;
+    r << host << ":" << port;
+    return r.str();
+  }
+  std::string node_id(const ReplicationNode& node) {
+    return node_id(node.host, node.port);
+  }
+
+  // Internal method to apply replication or set new value and replicate
+  void update(std::pair<std::string, uint32_t> tuple, bool replicate){
+    if(! is_ready){
+      throw std::logic_error("Replication is not ready");
+    }
+    state.MutableUse([tuple, replicate, this](SharedState& state) {
+      // Add local data
+      state.data[tuple.first] = tuple.second;
+
+      // Set clock and send update to replication thread
+      if(replicate) {
+        state.clock[tuple.first] = current::time::Now();
+        state.replication_out.push(std::move(tuple));
+      }
+    });
+  }
+
+ public:
+  AsyncReplicatedContainer(ReplicationConfig &config){
+      sid = node_id(config.host, config.port);
+      reader_port = config.port;
+      delay = config.delay;
+      nodes = config.nodes_list;
+      is_verbose = config.is_verbose;
+      show_network_err = config.show_network_errors;
+      max_waits = config.max_waits;
+  }
+  ~AsyncReplicatedContainer(){
+    stop();
+  }
+
+  void reader(uint16_t port){
+    int waits = 0;
+    bool is_stop = false;
+    Relay buffer;
+
+    while (true) {
+      is_stop = state.ImmutableUse(
+          [](SharedState const&state){ return state.die;});
+      if(is_stop){
+        break;
+      }
+      try {
+        current::net::Socket socket((current::net::BarePort(port)));
+        // FIXME? how to accept multiple connections on same port with SO_REUSEADDR/REUSEPORT
+        current::net::Connection connection(socket.Accept());
+
+        if(is_verbose) {
+          state.MutableUse([port](SharedState& state) {
+            std::cout << "Reader connected on port " << port << std::endl;
+          });
+        }
+
+        while (waits < max_waits) {
+          is_stop = state.ImmutableUse(
```

Comment:
Why do you name this variable at all, just curious/

Reply:
done